### PR TITLE
fix(controller): Sticky session correction for AWS ALB. Fixes #1572

### LINKED ISSUE #1572: Rollout clobbers pre-existing ALB annotations
## Summary 

We have pre-existing annotations in place for our ALB listener rules, but argo appears to be clobbering those annotations so they never actually take effect.

Our service requires the use of sticky sessions.  For enabling stickiness on targets, this is done by first setting this ingress annotation:

```yaml
alb.ingress.kubernetes.io/target-group-attributes: stickiness.enabled=true,stickiness.lb_cookie.duration_seconds=60
```

But since both the canary and stable target groups are routed by the same listener rule, we need to also enable target group stickiness at the rule level by adding setting the `TargetGroupStickinessConfig` in the following annotation (as [described here](https://aws.amazon.com/blogs/aws/new-application-load-balancer-simplifies-deployment-with-weighted-target-groups/)):

```yaml
alb.ingress.kubernetes.io/actions.xvp-ping-api-dev-stable:
  "{\"ForwardConfig\":{\"TargetGroupStickinessConfig\":{\"DurationSeconds\":300,\"Enabled\":true},\"TargetGroups\":[{\"ServiceName\":\"xvp-ping-api-dev-canary\",\"ServicePort\":8080,\"Weight\":50},{\"ServiceName\":\"xvp-ping-api-dev-stable\",\"ServicePort\":8080,\"Weight\":50}]},\"Type\":\"forward\"}",
```

Without specifying the `TargetGroupStickinessConfig` in our annotation, we see the following error from the `aws-load-balancer-controller` when it tries to update the ALB rules:

> **InvalidLoadBalancerAction**: You must enable group stickiness on a rule if you enabled target stickiness on one of its target groups

After some debugging, we can see in the controller logs that its trying to push the following payload to the ELB API:

```json
{
  "type": "forward",
  "forwardConfig": {
    "targetGroups": [
      {
        "targetGroupARN": { "$ref": "#/resources/AWS::ElasticLoadBalancingV2::TargetGroup/xvp-ping-api-dev/xvp-ping-api-dev-alb-xvp-ping-api-dev-stable:8080/status/targetGroupARN" },
        "weight": 100
      },
      {
        "targetGroupARN": { "$ref": "#/resources/AWS::ElasticLoadBalancingV2::TargetGroup/xvp-ping-api-dev/xvp-ping-api-dev-alb-xvp-ping-api-dev-canary:8080/status/targetGroupARN" },
        "weight": 0
      }
    ]
  }
}
```

We can see in the above output that the `TargetGroupStickinessConfig` configuration we set in our annotation is missing.  As it stands now, we're unable to enable sticky sessions with argo.

If I'm understanding the code [here](https://github.com/argoproj/argo-rollouts/blob/2565c87942d2236c171495a0cfdb6c1675fc3b67/ingress/alb.go#L85-L98) correctly, any pre-existing annotations are not considered when building the listener rule's `forwardConfig`.  This value is explicitly set with target group values and nothing else.

Should this code be updated to more gracefully handle annotations with pre-existing values so they are not lost?  Or perhaps a sort of template may be provided to use as a basis for building the ingress annotation?

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

### PR DESCRIPTION
**Pull request by derjust**
When sticky sessions are activated on the Ingress for AWS ALB:
```
Annotations:  
              alb.ingress.kubernetes.io/target-group-attributes: stickiness.enabled=true,stickiness.lb_cookie.duration_seconds=60
...
```
also targetGroup stickiness must be configured.
Otherwise the error
```
InvalidLoadBalancerAction: You must enable group stickiness on a rule if you enabled target stickiness on one of its target groups
```
is reported by the AWS ALB.

Therefore this PR adds support for AWS ALB [TargetGroupStickinessConfig](https://aws.amazon.com/blogs/aws/new-application-load-balancer-simplifies-deployment-with-weighted-target-groups/) following the same pattern as the `additionalIngressAnnotations`for Istio on the Rollout:

```
spec:
  progressDeadlineSeconds: 1800
  replicas: 2
  strategy:
    canary:
      canaryService: argo-test-canary
      stableService: argo-test-stable
      steps:
      - setWeight: 50
      - pause: {}
      - setWeight: 100
      trafficRouting:
        alb:
          ingress: argo-test-alb
          servicePort: 80
          stickinessConfig:
            durationSeconds: 60
            enabled: true
```

Screen Shot 2021-10-12 at 1 08 41 AM

---

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [X] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [X] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by harikrongali** [comment]
please fix CI failues

**Comment by derjust** [comment]
I was wondering how this PR can further proceed?

File: VERSION [hunk 1]
```diff
@@ -1 +1 @@
-1.1.0
+1.1.0-Sticky1
```
Comment (harikrongali) [inline VERSION:1]: please undo this change

**Comment by harikrongali** [comment]
@derjust We will review and get this sooner. Meantime, can you rebase to latest as CI checks are failing

File: rollout/trafficrouting/alb/alb.go [hunk 2]
```diff
@@ -243,6 +243,16 @@ func getForwardActionString(r *v1alpha1.Rollout, port int32, desiredWeight int32
 			TargetGroups: targetGroups,
 		},
 	}
+
+	var stickinessConfig = r.Spec.Strategy.Canary.TrafficRouting.ALB.StickinessConfig
+	if stickinessConfig.Enabled {
+		newStickyConfig := ingressutil.ALBTargetGroupStickinessConfig{
+			Enabled: true,
+			DurationSeconds: stickinessConfig.DurationSeconds,
```
[the remark this answers is not in the document]
Comment (harikrongali) [reply rollout/trafficrouting/alb/alb.go:241]: same range as specified by AWS API should be good

File: ingress/alb_test.go [hunk 3]
```diff
@@ -32,15 +32,42 @@ const actionTemplate = `{
...
 	return fmt.Sprintf("%s%s%s", ingressutil.ALBIngressAnnotation, ingressutil.ALBActionPrefix, stable)
 }
 
-func newALBIngress(name string, port int, serviceName string, rollout string) *extensionsv1beta1.Ingress {
+func newALBIngress(name string, port int, serviceName string, rollout string, includeStickyConfig bool) *extensionsv1beta1.Ingress {
 	canaryService := fmt.Sprintf("%s-canary", serviceName)
 	albActionKey := albActionAnnotation(serviceName)
 	managedBy := fmt.Sprintf("%s:%s", rollout, albActionKey)
-	action := fmt.Sprintf(actionTemplate, serviceName, port, canaryService, port)
+	var template string
```
Comment (harikrongali) [inline ingress/alb_test.go:64]: can we refactor to 
```
template := actionTemplate
if includeStickyConfig {
		template = actionTemplateWithStickyConfig
	}
```

File: rollout/trafficrouting/alb/alb_test.go [hunk 4]
```diff
@@ -67,15 +67,41 @@ const actionTemplate = `{
...
 
 func albActionAnnotation(stable string) string {
 	return fmt.Sprintf("%s%s%s", ingressutil.ALBIngressAnnotation, ingressutil.ALBActionPrefix, stable)
 }
 
-func ingress(name string, stableSvc, canarySvc string, port, weight int32, managedBy string) *extensionsv1beta1.Ingress {
+func ingress(name string, stableSvc, canarySvc string, port, weight int32, managedBy string, includeStickyConfig bool) *extensionsv1beta1.Ingress {
 	managedByValue := fmt.Sprintf("%s:%s", managedBy, albActionAnnotation(stableSvc))
-	action := fmt.Sprintf(actionTemplate, canarySvc, port, weight, stableSvc, port, 100-weight)
+	var action string
```
Comment (harikrongali) [inline rollout/trafficrouting/alb/alb_test.go:99]: refactor to
```
action := fmt.Sprintf(actionTemplate, canarySvc, port, weight, stableSvc, port, 100-weight)
if includeStickyConfig {
		action = fmt.Sprintf(actionTemplateWithStickyConfig, canarySvc, port, weight, stableSvc, port, 100-weight)
}
```

**Comment by harikrongali** [comment]
@derjust overall looks good. Thanks for the contribution. As soon as the comments & CI failures are fixed, I will get this merged

**Comment by harikrongali** [comment]
Please run codegen locally and commit changes generated from it.

**Comment by derjust** [comment]
Thank you for the extensive review - it's my first Go PR so I appreciate all the effort.
Working on this now

**Comment by harikrongali** [comment]
@derjust there are some unit tests failing with nil pointer exceptions.
```FAIL: TestGetForwardActionStringMarshalsDisabledStickyConfigCorrectly (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1993061]
```

can you run "make test" locally?

File: pkg/signals/signal_posix.go [hunk 5]
```diff
@@ -1,3 +1,4 @@
+//go:build !windows
```
[the remark this answers is not in the document]
Comment (harikrongali) [reply pkg/signals/signal_posix.go:1]: even the comment part as well..so that it wont appear in diff

Comment (harikrongali) [inline pkg/signals/signal_posix.go:1]: this is causing codegen CI to fail..can you undo change to this file

Comment (derjust) [reply pkg/signals/signal_posix.go:1]: Not sure where this came in. Removed

**Comment by harikrongali** [comment]
@derjust please do steps mentioned in for DCO Check https://github.com/argoproj/argo-rollouts/pull/1577/checks?check_run_id=4268869529

File: pkg/signals/signal_posix.go (hunk 5 above)
Comment (derjust) [reply pkg/signals/signal_posix.go]: Figured it out: the `make codegen` created this change. I removed if

File: pkg/apis/rollouts/v1alpha1/zz_generated.deepcopy.go [hunk 6]
```diff
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
```
Comment (harikrongali) [inline pkg/apis/rollouts/v1alpha1/zz_generated.deepcopy.go:1]: code-gen failing for this line. please remove this line

Comment (derjust) [reply pkg/apis/rollouts/v1alpha1/zz_generated.deepcopy.go]: Thank you; undone.
Does the `code-gen` has an issue running on Mac?

**Comment by derjust** [comment]
Unit tests work locally (Note: they fail on OSX with a `permission denied` - which makes the `list` test fail as the clock (I guess) can't be reset.
DCO also fixed. Thank you for the ongoing support!

**Comment by harikrongali** [comment]
@derjust seems the code-gen that you are running has issues. can you let me know if you followed https://argoproj.github.io/argo-rollouts/CONTRIBUTING/
installation
```
brew install go kubectl kustomize golangci-lint protobuf swagger-codegen
```
make codegen - Runs the code generator that creates the informers, client, lister, and deepcopies from the types.go and modifies the open-api spec.

**Comment by derjust** [comment]
@harikrongali  thank you for your continued help.
Yes i did so - ran it again, too:
```
$ brew install go kubectl kustomize golangci-lint protobuf swagger-codegen

Warning: go 1.17.2 is already installed and up-to-date.
To reinstall 1.17.2, run:
  brew reinstall go
Warning: kubernetes-cli 1.22.4 is already installed, it's just not linked.
To link this version, run:
  brew link kubernetes-cli
Warning: kustomize 4.4.1 is already installed and up-to-date.
To reinstall 4.4.1, run:
  brew reinstall kustomize
Warning: golangci-lint 1.43.0 is already installed and up-to-date.
To reinstall 1.43.0, run:
  brew reinstall golangci-lint
Warning: protobuf 3.17.3 is already installed and up-to-date.
To reinstall 3.17.3, run:
  brew reinstall protobuf
Warning: swagger-codegen 3.0.29 is already installed and up-to-date.
To reinstall 3.0.29, run:
  brew reinstall swagger-codegen
```
To help it i force-use GoLang 1.16 - for some reason `TestFunctionalSuite/TestWorkloadRef` still fails :-(

**Comment by derjust** [comment]
Did more investigation and posted my findings around my codegen issue - and failing e2e test - as new issue: 
https://github.com/argoproj/argo-rollouts/issues/1675

Hope this can shed some light?

**Comment by derjust** [comment]
A PR for the issue in  #1675  was provided.
As it is already approved i assume it will be merged to `master` soon - so i already re-based this PR on it already.

**Comment by harikrongali** [comment]
@derjust sorry, I am out the last couple of weeks. I will get this merged this week.

**Review by harikrongali — APPROVED** [review]

**Comment by harikrongali** [comment]
@alexmt can you merge the PR. e2e test failed is a flaky one not related the current change.

**Review by alexmt — APPROVED** [review]

**Pull request merged**